sys/kern/signal.c

```diff
@@ -342,15 +347,27 @@ void sig_kill(proc_t *p, ksiginfo_t *ksi) {
     return;
   }
 
+  sigprop_t prop = sig_properties[sig];
+
+  /* Don't send TTY stop signals to members of an orphaned process group
+   * with a default handler.
+   * Theoretically, we should hold all_proc_mtx for reading pg_jobc,
+   * but the race here is harmless. */
+  if ((prop & SA_TTYSTOP) && (p->p_pgrp->pg_jobc == 0) && (handler == SIG_DFL))
```

Comment:
I don't understand why we can do this without holding `all_proc_mtx`. Could you explain more?

Reply:
Actually, when writing my response I realized that this race can result in unwanted behavior: a process in an orphaned process group stopping on a terminal stop signal. However, the second race (the one in `sig_post()`) is actually harmless, here's why:

In order for a process in an orphaned pgrp to stop on a terminal stop signal, it must observe in `sig_post()` that its `jobc` is greater than 0. Suppose that immediately after reading `jobc`, the process group becomes orphaned. In `pgrp_maybe_orphan()`, we iterate over all processes in the orphaned group and send `SIGHUP` and `SIGCONT` to all stopped processes. The process state check is done while holding the process's `p_lock`, which `sig_post()` also holds. So, `pgrp_maybe_orphan()` won't examine the state of the process until `sig_post()` releases the lock, which it does __after__ stopping the process. So, when `pgrp_maybe_orphan()` acquires the lock, it will see that the process is stopped, and it will send `SIGCONT` to it, which is what we want to happen.